Add Firewall resource

psinghal20 opened the pull request:
### Description of your changes

Fixes a part of #329 

Some of the things I have assumed while defining types and controller that can be incorrect and would require a look:
- The fields marked as `OUTPUT_ONLY` are to be treated as observations.
- The fields `ForceSendFields`, `NullFields` and `ServerResponse` in google API types can be ignored for our use case.
- All the comments in Type fields are directly used from google API and were a little verbose.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested
This code has primarily been tested locally using the example file attached with this change. I have also added unit tests similar to what we have for other compute resources.

hasheddan commented on pkg/controller/gcp.go:
```diff
@@ -43,7 +43,9 @@ func Setup(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter) error {
 		compute.SetupSubnetwork,
+		compute.SetupFirewall,
 		container.SetupCluster,
+		container.SetupGKECluster,
```
I believe this may have been inadvertently added during rebase.

psinghal20 replied:
Oh, I think I messed up while rebasing the master to fix the conflicts, I will fix this up, and I think ideally it should just be `container.SetupCluster`. My bad, I will fix this up.

hasheddan commented on apis/compute/v1beta1/firewall_types.go:
```diff
@@ -0,0 +1,265 @@
+type FirewallLogConfig struct {
+	// Enable: This field denotes whether to enable logging for a particular
+	// firewall rule.
+	Enable bool `json:"enable,omitempty"`
```
If this is optional it should be `*bool` and marked with `+ optional`. However, sometimes when there is a single field in an embedded struct I will opt for making it required since the embedded struct really shouldn't be provided if the only field it supports is not set.

psinghal20 replied:
Right, I agree with that. I have made `FirewallLogConfig` an optional object but made `Enable` as required and removed `omitempty` for it.

hasheddan commented on apis/compute/v1beta1/firewall_types.go:
```diff
@@ -0,0 +1,265 @@
+	// type is required when creating a firewall rule. This value can either
+	// be one of the following well known protocol strings (tcp, udp, icmp,
+	// esp, ah, ipip, sctp) or the IP protocol number.
+	IPProtocol string `json:"IPProtocol,omitempty"`
```
Should this be required within the embedded struct? If so it should have `omitempty` removed.

hasheddan commented on apis/compute/v1beta1/firewall_types.go:
```diff
@@ -0,0 +1,265 @@
+	// type is required when creating a firewall rule. This value can either
+	// be one of the following well known protocol strings (tcp, udp, icmp,
+	// esp, ah, ipip, sctp) or the IP protocol number.
+	IPProtocol string `json:"IPProtocol,omitempty"`
```
Should this be required within the embedded struct? If so it should have `omitempty` removed.

hasheddan reviewed (commented):
This is looking really great @psinghal20! A few comments below and it would be great to also add some unit tests for the firewall client :)

psinghal20 commented:
Hi @hasheddan, I have resolved all the above comments and added some small unit tests for firewall client code which helped uncover an issue in the `generateFirewall` method. :grimacing:

hasheddan reviewed (commented):
@psinghal20 was revisiting this PR and I noticed this resource is being introduced at `v1beta1`, but we typically introduced new resources at `v1alpha1`, would you mind updating?

turkenh commented on pkg/controller/compute/firewall.go:
```diff
@@ -0,0 +1,184 @@
+		return managed.ExternalUpdate{}, errors.New(errNotFirewall)
+	}
+
+	observed, err := c.Firewalls.Get(c.projectID, meta.GetExternalName(cr)).Context(ctx).Do()
+	if err != nil {
+		return managed.ExternalUpdate{}, errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errGetFirewall)
+	}
+
+	upToDate, err := firewall.IsUpToDate(meta.GetExternalName(cr), &cr.Spec.ForProvider, observed)
+	if err != nil {
+		return managed.ExternalUpdate{}, errors.Wrap(err, errCheckFirewallUpToDate)
+	}
+	if upToDate {
+		return managed.ExternalUpdate{}, nil
+	}
```
I believe we already did our observation in `Observe` above. Any reason to make another observation here?

turkenh commented on pkg/clients/firewall/firewall.go:
```diff
@@ -0,0 +1,139 @@
+		return true, errors.New(errCheckUpToDate)
+	}
+	GenerateFirewall(name, *in, desired)
+	return cmp.Equal(desired, observed, cmpopts.EquateEmpty(), cmpopts.IgnoreFields(compute.Firewall{}, "ForceSendFields")), nil
```
While testing on my side, I noticed `IsUpToDate` keeps returning `false`. To debug further, I've added following lines just above return:

```
	diff := cmp.Diff(desired, observed, cmpopts.EquateEmpty(), cmpopts.IgnoreFields(compute.Firewall{}, "ForceSendFields"))
	fmt.Println(diff)
```

Which printed:
```
  &compute.Firewall{
        ... // 9 identical fields
        LogConfig: &{},
        Name:      "firewall-example-by-hasan",
-       Network:   "projects/wesaas-playground/global/networks/example-by-hasan",
+       Network:   "https://www.googleapis.com/compute/v1/projects/wesaas-playground/global/networks/example-by-hasan",
        Priority:  1000,
        SelfLink:  "https://www.googleapis.com/compute/v1/projects/wesaas-playground"...,
        ... // 1 ignored and 7 identical fields
  }
```

According [GCP API reference](https://cloud.google.com/compute/docs/reference/rest/v1/firewalls), both are valid for network field, however, we count this as a diff and keep sending updates to the resource.

turkenh reviewed (commented):
@psinghal20, first of all, sorry for the long review process but I hope there will be some improvement in the near future.
Your PR looking very good and thanks a lot for your contribution 💪

While testing on my side, I observed some missing pieces which I added as comments.

> @psinghal20 was revisiting this PR and I noticed this resource is being introduced at `v1beta1`, but we typically introduced new resources at `v1alpha1`, would you mind updating?

I also noticed you already moved to v1alpha1 but didn't register to schema which causes controller to fail at startup. Basically, we just need the following patch:

```
diff --git a/apis/gcp.go b/apis/gcp.go
index 14c59b3..269efbe 100644
--- a/apis/gcp.go
+++ b/apis/gcp.go
@@ -21,6 +21,7 @@ import (
        "k8s.io/apimachinery/pkg/runtime"

        cachev1beta1 "github.com/crossplane/provider-gcp/apis/cache/v1beta1"
+       computev1alpha1 "github.com/crossplane/provider-gcp/apis/compute/v1alpha1"
        computev1beta1 "github.com/crossplane/provider-gcp/apis/compute/v1beta1"
        containerv1beta1 "github.com/crossplane/provider-gcp/apis/container/v1beta1"
        containerv1beta2 "github.com/crossplane/provider-gcp/apis/container/v1beta2"
@@ -41,6 +42,7 @@ func init() {
                gcpv1alpha3.SchemeBuilder.AddToScheme,
                gcpv1beta1.SchemeBuilder.AddToScheme,
                cachev1beta1.SchemeBuilder.AddToScheme,
+               computev1alpha1.SchemeBuilder.AddToScheme,
                computev1beta1.SchemeBuilder.AddToScheme,
                containerv1beta2.SchemeBuilder.AddToScheme,
                containerv1beta1.SchemeBuilder.AddToScheme,
```

psinghal20 commented:
Hi @turkenh, I have resolved the above-mentioned comments. Please take a look, thanks!

turkenh commented:
@psinghal20 thanks, looking good but there seem to be some linter errors. 
Could you check them?

fahedouch commented:
Hi teams, 

thank you for your work @psinghal20 . @hasheddan  @turkenh we really need this feature, any idea in which release it will land?

psinghal20 commented:
Hi @turkenh, I have updated the PR, the change required a rebase with the master. I have updated some of the tests accordingly, and removed the test case checking the Kubernetes resource update which we removed [here](https://github.com/crossplane/provider-gcp/pull/333#discussion_r692034730). Let me know if anything else is required, thanks! :smile:

turkenh reviewed (approved):
Great work @psinghal20, thanks!
